rslib/src/import_export/text/import.rs

```diff
@@ -0,0 +1,216 @@
+        Ok(Self {
+            col,
+            usn,
+            normalize_notes,
+            notetypes,
+            deck_ids,
+        })
+    }
+
+    fn import_foreign_notetypes(&mut self, notetypes: Vec<ForeignNotetype>) -> Result<()> {
+        for foreign in notetypes {
+            let mut notetype = foreign.into_native();
+            notetype.usn = self.usn;
+            self.col
+                .add_notetype_inner(&mut notetype, self.usn, false)?;
+        }
+        Ok(())
+    }
+    fn notetype_for_note(&mut self, note: &ForeignNote) -> Result<Option<Arc<Notetype>>> {
+        Ok(if let Some(nt) = self.notetypes.get(&note.notetype) {
+            nt.clone()
+        } else {
+            let nt = self.col.notetype_for_string(&note.notetype)?;
+            self.notetypes.insert(note.notetype.clone(), nt.clone());
+            nt
+        })
+    }
+
+    fn deck_id_for_note(&mut self, note: &ForeignNote) -> Result<Option<DeckId>> {
+        Ok(if let Some(did) = self.deck_ids.get(&note.deck) {
+            *did
+        } else {
+            let did = self.col.deck_id_for_string(&note.deck)?;
+            self.deck_ids.insert(note.deck.clone(), did);
+            did
+        })
+    }
+
+    fn import_foreign_notes(&mut self, notes: Vec<ForeignNote>) -> Result<NoteLog> {
+        let mut log = NoteLog::default();
+        for foreign in notes {
+            if let Some(notetype) = self.notetype_for_note(&foreign)? {
+                if let Some(deck_id) = self.deck_id_for_note(&foreign)? {
+                    let log_note = self.import_foreign_note(foreign, &notetype, deck_id)?;
+                    log.new.push(log_note);
+                }
+            }
+        }
+        Ok(log)
+    }
+
+    fn import_foreign_note(
+        &mut self,
+        foreign: ForeignNote,
+        notetype: &Notetype,
+        deck_id: DeckId,
+    ) -> Result<LogNote> {
+        let (mut note, mut cards) = foreign.into_native(notetype, deck_id);
+        self.import_note(&mut note, notetype)?;
+        self.import_cards(&mut cards, note.id)?;
+        self.generate_missing_cards(notetype, deck_id, &note)?;
+        Ok(note.into_log_note())
+    }
+
+    fn import_note(&mut self, note: &mut Note, notetype: &Notetype) -> Result<()> {
+        self.col.canonify_note_tags(note, self.usn)?;
+        note.prepare_for_update(notetype, self.normalize_notes)?;
+        note.usn = self.usn;
+        self.col.add_note_only_undoable(note)
+    }
+
+    fn import_cards(&mut self, cards: &mut [Card], note_id: NoteId) -> Result<()> {
+        for card in cards {
+            card.note_id = note_id;
+            self.col.add_card(card)?;
+        }
+        Ok(())
+    }
+
+    fn generate_missing_cards(
+        &mut self,
+        notetype: &Notetype,
+        deck_id: DeckId,
+        note: &Note,
+    ) -> Result<()> {
+        let card_gen_context = CardGenContext::new(notetype, Some(deck_id), self.usn);
```

Comment:
This parses all the templates in the notetype, so ideally we don't want to be doing this for every note.